Sweetener.Logging/LogPattern.cs

```diff
@@ -0,0 +1,385 @@
+                }
+
+                return null;
+            }
+
+            private static Parameter ParseParameter(ref int i, string pattern)
+            {
+                // Need to keep track of the number of characters read because it is not
+                // necessarily the same as the size of the name. Spaces are allowed before or after.
+                int nameStart = -1, nameEnd = -1;
+                for (; i < pattern.Length; i++)
+                {
+                    char c = pattern[i];
+                    // There are currently no patterns with numbers enabled
+                    if (char.IsLetter(c) && nameEnd == -1)
+                    {
+                        if (nameStart == -1)
+                            nameStart = i;
+                    }
+                    else if (c == ' ')
+                    {
+                        if (nameStart != -1 && nameEnd == -1)
+                            nameEnd = i;
+                    }
+                    else if (c == ',' || c == ':' || c == '}')
+                    {
+                        if (nameEnd == -1)
+                            nameEnd = i;
+
+                        break;
+                    }
+                    else
+                    {
+                        throw new FormatException("Input string was not in a correct format");
+                    }
+                }
+
+                // Did we reach the end without breaking?
+                if (i >= pattern.Length)
+                    throw new FormatException("Input string was not in a correct format");
+
+                string name = pattern.Substring(nameStart, nameEnd - nameStart);
+
+                // TODO: Case insensitive?
+                switch (name)
+                {
+                    case "msg":
+                    case "message":
+                        return Parameter.Message; 
+                    case "ts":
+                    case "timestamp":
+                        return Parameter.Timestamp; 
+                    case "l":
+                    case "level":
+                        return Parameter.Level;
+                    case "pid":
+                    case "processId":
+                        return Parameter.ProcessId;
+                    case "pn":
+                    case "processName":
+                        return Parameter.ProcessName;
+                    case "tid":
+                    case "threadId":
+                        return Parameter.ThreadId;
+                    case "tn":
+                    case "threadName":
+                        return Parameter.ThreadName;
+                    default:
+                        throw new FormatException($"Unrecognized pattern variable '{name}'.");
+                }
+            }
+        }
+
+        /// <summary>
+        /// A known variable in the log layout pattern.
+        /// </summary>
+        public enum Parameter
```

Comment:
SInce this is a public enum  does it have to live in this class? There is a lot going on in this class, consider pulling it out into its own file like LogLevel.

Reply:
It actually isn’t public because the encompassing class is internal. I am unsure if I want to expose it publicly. If we make it public, it should probably live in its own file.

Do you think it’s worth exposing? I don’t think there is a use case for it. Perhaps there could be a set of constants for the pattern parameter names if necessary.

Reply:
Forgot the class is internal. No need to make this public.